libzerocash/Coin.cpp

```diff
@@ -22,6 +22,48 @@ Coin::Coin(): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
 
 }
 
+Coin::Coin(const std::string bucket, Address& addr): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
+    // Retreive and decode the private key
+    ECIES<ECP>::PrivateKey decodedPrivateKey;
+    decodedPrivateKey.Load(StringStore(addr.getEncryptionSecretKey()).Ref());
+
+    // Create the decryption session
+    AutoSeededRandomPool prng;
+    ECIES<ECP>::Decryptor decrypt(decodedPrivateKey);
+
+    // Convert the input string into a vector of bytes
+    std::vector<byte> bucket_bytes(bucket.begin(), bucket.end());
+
+    // Construct a temporary object to store the plaintext, large enough
+    // to store the plaintext if it were extended beyond the real size.
+    std::vector<unsigned char> plaintext;
+    // Size as needed, filling with zeros.
+    plaintext.resize(decrypt.MaxPlaintextLength(decrypt.CiphertextLength(v_size + zc_r_size + rho_size)), 0);
```

Comment:
@ebfull @nathan-at-least The subexpression ``decrypt.CiphertextLength(v_size + zc_r_size + rho_size)`` is repeated here and in the call to ``decrypt.Decrypt``. It would be clearer to store that in a temporary ``ciphertext_length`` variable.